### ISSUE TITLE
perf(direct): torch.compile RAUNE-Net for Inductor JIT fusion

### ISSUE THREAD

**Issue by chunzhe10** [chunzhe10/dorea#69]
## Problem

After fp16 landed (chunzhe10/dorea#67), direct mode is at 4.25 fps on 4K 120fps HEVC, with per-stage timing:

\`\`\`
decode:  34 ms/frame (cheap, hidden behind GPU)
gpu:    230 ms/frame ← critical path, dominated by RAUNE forward pass
encode:  50-100 ms/frame (cheap, hidden behind GPU)
wall:   235 ms/frame ≈ gpu_busy (pipeline perfectly saturated)
```

Issue #68 (batched uploads) was closed after proving that PCIe overhead is NOT the bottleneck — CUDA streams already overlap H2D transfers with GPU compute. The actual bottleneck is RAUNE-Net forward-pass compute itself.

## Proposed fix

Wrap the fp16 RAUNE model with `torch.compile()` to get Inductor JIT fusion of the 30 residual blocks. PyTorch 2.x Inductor can fuse element-wise ops (conv → batchnorm → relu → skip-add), eliminating intermediate tensor materialization and reducing kernel launch count.

\`\`\`python
model = model.half()
for m in model.modules():
    if isinstance(m, (nn.InstanceNorm2d, ...)):
        m.float()
# NEW:
model = torch.compile(model, mode="reduce-overhead")
\`\`\`

## Expected outcome

- **~1.2-1.5× speedup on RAUNE forward pass** (industry benchmark for vision models on Ampere)
- 4.25 fps → ~5-6 fps target
- No quality regression (compile produces numerically equivalent output within fp16 rounding)
- **Cold-start cost**: first call triggers Inductor compilation (~10-30s). Subsequent calls cached via `TORCHINDUCTOR_CACHE_DIR`.

## Constraints

- Quality preserving (numerically equivalent to non-compiled)
- VRAM must stay under 6 GB
- Cold-start delay is acceptable for offline batch processing but should be logged
- Must be compatible with the InstanceNorm2d fp32 precision fix from #67

## Risks and mitigations

### Risk: torch.compile has edge cases with certain layer types
**Mitigation**: Test on the POC benchmark first. If compile fails, fall back to eager.

### Risk: Compilation cache invalidation
**Mitigation**: Use persistent cache via env var. Log when recompilation occurs.

### Risk: Output divergence from non-compiled
**Mitigation**: Frame parity check vs current baseline (target YMAX ≤ 3/255).

## Reference

- #66 (parent GPU optimization sweep — PR 2 of 3)
- #67 (fp16 baseline this builds on)
- #68 closed wontfix (proved upload batching was not the bottleneck)
- Bottleneck analysis corvia entries: 019d74c9, 019d753f

**Comment by chunzhe10** [comment]
🤖 Claimed by Claude Code agent — starting dev-loop.

**Comment by chunzhe10** [comment]
## Closing as wontfix — torch.compile regresses on this workload

Implemented and benchmarked torch.compile on branch `feat/69-torch-compile` (never pushed). Results show it actively regresses performance AND introduces visual differences above tolerance.

### Measured results

| Config | Throughput | GPU busy | Visual diff (YMAX) |
|---|---|---|---|
| fp16 baseline (#67) | 4.32 fps | 227.7 ms | 0 (bit-exact) |
| **torch.compile reduce-overhead** | **3.90 fps** | **252.4 ms** | **13 (over threshold)** |

- Throughput: -10%
- GPU busy: +11%
- Visual diff YMAX: 13/255 vs tolerance of 5/255

### Root causes

1. **Not enough SMs on RTX 3060.** Inductor logged `"Not enough SMs to use max_autotune_gemm mode"`. The RTX 3060 has 28 SMs. Inductor's tile autotuning is designed for A100-class GPUs (108 SMs). On a 28-SM consumer card, Inductor's generated kernels don't match what cuDNN already ships.

2. **cuDNN eager mode is already well-tuned** for RAUNE-Net's 30 residual blocks on consumer Ampere. There's no slack for Inductor to recover.

3. **CUDA graph thread-locality deadlock.** `mode="reduce-overhead"` captures CUDA graphs which are thread-local. Capturing on the main thread (warmup) and replaying from the GPU worker thread deadlocks. Working around this required removing the warmup, which means the capture happens on the first real batch instead — negating the main benefit of `reduce-overhead` mode.

4. **fp16 rounding divergence.** Inductor's fused kernels produce slightly different fp16 rounding than the eager cuDNN path. Worst-case pixel diff hits 13/255, above the visual tolerance threshold.

### Pattern: the GPU optimization sweep hit diminishing returns

Two "GPU stage optimization" issues closed as wontfix now:
- #68 batched uploads: wrong premise — PCIe wasn't the bottleneck
- #69 torch.compile: wrong hardware — Inductor doesn't optimize for 28-SM consumer GPUs

fp16 (#67) extracted the low-hanging fruit. The remaining ~230 ms/frame GPU time is cuDNN-optimal RAUNE-Net forward pass, and there's no more free speedup without changing the model itself, the proxy resolution (removed from scope per user constraint), or frame skipping (also removed from scope).

### What's left in issue #66 (the parent sweep)

**PR 3/3 was NVDEC hardware decode.** But per-stage profile shows decode is only ~34 ms/frame and already hidden behind GPU work in the 3-thread pipeline. Even eliminating decode entirely saves no measurable wall-clock fps, because the slowest stage determines throughput.

### Practical conclusion

**4.25 fps is the practical ceiling for direct mode on RTX 3060 + 6GB VRAM under the "no quality loss" constraint.** To go faster requires lifting the quality constraint (lower proxy resolution, frame skipping) OR upgrading hardware.

Closing this issue. Will likely also close #66 (the parent sweep) once the NVDEC question is settled.

Branch `feat/69-torch-compile` was never pushed; deleted locally.
